### DISCUSSION TITLE
How can I customize the 404 page?

### DISCUSSION THREAD

**Discussion by mickmcg** [facebook/docusaurus#6030] [Q&A] [answered]
I initially tried to do this in Netlify with a _redirects file to point to my 404.html for any URLs not found. With that method, you see the custom 404 page flash up for a split second before the standard Docusaurus page is shown instead with this text.

Page Not Found
We could not find what you were looking for.
Please contact the owner of the site that linked you to the original URL and let them know their link is broken.

Is there a proper Docusaurus way to have custom 404 pages?

**Comment by Josh-Cena** [comment, marked as answer]
Hi, you can swizzle the `NotFound` component. https://docusaurus.io/docs/using-themes#swizzling-theme-components

And ignore any warnings / cautions that might come up. This is a very safe component.

**Reply by plmercereau** [reply]
The above link goes to a 404 - oh the irony 😊
The correct link to the [swizzling page](https://docusaurus.io/docs/swizzling)

**Reply by Josh-Cena** [reply]
Thanks for the report @plmercereau I'll add a redirect

**Reply by B4nan** [reply]
Looks like this is not working anymore in v2.0.1 stable, I can swizzle the component just fine, but no changes to it are reflected? Or is this taken into account only for production builds? I was testing this only locally so far.

**Reply by Josh-Cena** [reply]
@B4nan It works for me locally. Did you restart your dev server?

**Reply by B4nan** [reply]
Ok my bad, the confusion coming from me changing the texts inside `<Translate>` tags, as that has no effect (which is quite weird to me, I would expect unmatched translation for the block instead, and seeing the text I wrote).

**Reply by B4nan** [reply]
btw the component is not listed at all when doing `npm run swizzle -- --list` (but when swizzling you dont get any warnings, just like with the ones considered stable)

**Reply by Josh-Cena** [reply]
Really? Shows for me:

<img width="1171" alt="image" src="https://user-images.githubusercontent.com/55398995/183610550-ee826699-cc85-4d10-a192-f1accb844030.png">

**Reply by B4nan** [reply]
Ok, again my bad, I was trying to search just `notfound` and was in a case sensitive mode 🤦 

Thanks for the prompt responses.

**Comment by mickmcg** [comment]
Hi Josh, thanks for the fast response. I can confirm this method works. 👍 

`npm run swizzle @docusaurus/theme-classic NotFound`

Then edit `src/theme/NotFound.js`

**Reply by andrewm4894** [reply]
Just adding an example for any newbies like me who are a little unsure about all this (as i was). 

I used this to eject and put in some code like this to fire a PostHog telemetry event each time a user hits the Not Found page. 

![image](https://user-images.githubusercontent.com/2178292/195601511-b212e463-1db4-42cc-9047-9654bc391d6a.png)

**Reply by slorber** [reply]
I'd suggest to --wrap instead of --eject for this use-case: you'd have less duplicated code to maintain on your side

**Reply by andrewm4894** [reply]
cool - good to know - i'll give that a go.

**Reply by kazisean** [reply]
> Hi Josh, thanks for the fast response. I can confirm this method works. +1
> 
> `npm run swizzle @docusaurus/theme-classic NotFound`
> 
> Then edit `src/theme/NotFound.js`

for any newbies like me, here is the default  NotFound.js file (From [andrewm4894](https://github.com/andrewm4894)) 

```
import React, { useEffect } from 'react';
import Translate, {translate} from '@docusaurus/Translate';
import {PageMetadata} from '@docusaurus/theme-common';
import Layout from '@theme/Layout';
export default function NotFound() {

  return (
    <>
      <PageMetadata
        title={translate({
          id: 'theme.NotFound.title',
          message: 'Page Not Found',
        })}
      />
      <Layout>
        <main className="container margin-vert--xl">
          <div className="row">
            <div className="col col--6 col--offset-3">
              <h1 className="hero__title">
                <Translate
                  id="theme.NotFound.title"
                  description="The title of the 404 page">
                  Page Not Found
                </Translate>
              </h1>
              <p>
                <Translate
                  id="theme.NotFound.p1"
                  description="The first paragraph of the 404 page">
                  We could adadadnot find what you were looking for.
                </Translate>
              </p>
              <p>
                <Translate
                  id="theme.NotFound.p2"
                  description="The 2nd paragraph of the 404 page">
                  Please contact the owner of the site that linked you to the
                  original URL and let them know their link is broken.
                </Translate>
              </p>
            </div>
          </div>
        </main>
      </Layout>
    </>
  );
}
```

**Comment by tisonkun** [comment]
Is it possible to provide a JSON return conditionally instead of always a page?

**Reply by slorber** [reply]
I don't know what you mean here 😅 Why would you want it to return a json and what conditionally means

**Reply by tisonkun** [reply]
Aha...

After a second thought, it seems reasonable to always return a 404 page if not found and I can distinguish the case by matching page content.

The original problem is:

```ts
    const { data } = useSWR('/api/auth/me', fetcher, {
        shouldRetryOnError: (error) => {
            // specially handle that Docusaurus returns a 404 page
            return error.status != 404 && !`${error}`.startsWith('SyntaxError')
        }
    })
```

Since the 404 page is returned with Docu, the status is 200 instead of 404 so I have to match the error content for specially handling.

**Reply by tisonkun** [reply]
Even I try to fake a `/api/auth/me` to return JSON content, I don't find a way to do so.

**Reply by slorber** [reply]
I still have no idea what you are trying to do sorry. Please start by explaining your goal instead of roadblocks.

> Since the 404 page is returned with Docu, the status is 200 instead of 404 so I have to match the error content for specially handling.

It is the responsibility of your host to return the status code on the initial request. Only the initial request has a status code. If you "navigate" to an unexisting page (with a broken link for example), there's no status code involved.

Your host can definitively return a 404 status code. Most hosts would do that by default.

![CleanShot 2022-12-28 at 10 19 14@2x](https://user-images.githubusercontent.com/749374/209789077-162c8c1e-3bac-4b06-857e-35c7badc11ce.png)

If you want `/api/auth/me` to return a JSON instead of a 404, again you can try adding an edge middleware for that 🤪 or probably just a regular node/serverless function too.

**Reply by tisonkun** [reply]
Thank you! This is not a roadblock to me anymore. I'm fine with all things provided now :)

**Reply by tisonkun** [reply]
@slorber For the status code part, here is a repo:

```bash
yarn start # with a docusaurus quickstart project
```

```bash
curl -i localhost:3000/api/auth/me # or any path not exist
```

You will get a 200 return:

```
curl -i localhost:3000/api/auth/me
HTTP/1.1 200 OK
X-Powered-By: Express
access-control-allow-origin: *
Content-Type: text/html; charset=utf-8
Accept-Ranges: bytes
Content-Length: 1285
ETag: W/"505-fYqcMyKlu4OHsJA9G4TaGdStWAg"
Vary: Accept-Encoding
Date: Wed, 28 Dec 2022 09:35:14 GMT
Connection: keep-alive
Keep-Alive: timeout=5

<!DOCTYPE html>
<html lang="en">
...
```

This is not an issue or something that asks for help. Just share a case. It doesn't require a feedback, also.

**Reply by slorber** [reply]
`yarn start` is a dev server, not a production host

You get a 404 on a production host (or you should configure it if you don't)

`curl -i https://docusaurus.io/api/auth/me`

<img width="492" alt="CleanShot 2022-12-28 at 11 57 37@2x" src="https://user-images.githubusercontent.com/749374/209801744-6b6aba60-4a6d-4c33-bd19-ed8405f762b8.png">

**Comment by mickmcg** [comment]
I just realized we had lost our custom 404 page, possibly since upgrading to Docuraurus 3.x

I tried the same steps as previously `npm run swizzle @docusaurus/theme-classic NotFound` and saw 2 options JavaScipt/TypeScript so I said JavaScript and then 2 options after that for swizzle action Wrap/Eject and I tried both when I customise the resulting `/src/theme/NotFound/Content/index.js` file I don't see my changes picked up.

![SCR-20240711-mhfw-2](https://github.com/facebook/docusaurus/assets/2614604/127b2fd5-414f-45ba-91ce-a31170e82741)

**Reply by mickmcg** [reply]
Please ignore my message above, it is working now. Don't know why it suddenly started working, maybe restarting the application although pretty sure I already did that more than once before posting so one of life's unexplained mysteries 😄

**Reply by slorber** [reply]
This is probably due to the fact that swizzle does not support hot reload properly (https://github.com/facebook/docusaurus/issues/10238)
